c++/src/io/Cache.cc

```diff
@@ -0,0 +1,172 @@
+    uint64_t coalescedEnd = coalescedStart + itr->length;
+
+    for (++itr; itr < ranges.end(); ++itr) {
+      const uint64_t currentRangeStart = itr->offset;
+      const uint64_t currentRangeEnd = currentRangeStart + itr->length;
+
+      assert(coalescedStart < coalescedEnd);
+      assert(currentRangeStart < currentRangeEnd);
+
+      // At this point, the coalesced range is [coalesced_start, prev_range_end).
+      // Stop coalescing if:
+      //   - coalesced range is too large, or
+      //   - distance (hole/gap) between consecutive ranges is too large.
+      if ((currentRangeEnd - coalescedStart > rangeSizeLimit) ||
+          (currentRangeStart > coalescedEnd + holeSizeLimit)) {
+        coalesced.push_back({coalescedStart, coalescedEnd - coalescedStart});
+        coalescedStart = currentRangeStart;
+      }
+
+      // Update the prev_range_end with the current range.
+      coalescedEnd = currentRangeEnd;
+    }
+    coalesced.push_back({coalescedStart, coalescedEnd - coalescedStart});
+
+    assert(coalesced.front().offset == ranges.front().offset);
+    assert(coalesced.back().offset + coalesced.back().length ==
+           ranges.back().offset + ranges.back().length);
+    return coalesced;
+  }
+
+  std::vector<ReadRange> coalesceReadRanges(std::vector<ReadRange> ranges, uint64_t holeSizeLimit,
+                                            uint64_t rangeSizeLimit) {
+    assert(rangeSizeLimit > holeSizeLimit);
+
+    ReadRangeCombiner combiner{holeSizeLimit, rangeSizeLimit};
+    return combiner.coalesce(std::move(ranges));
+  }
+
+  void ReadRangeCache::cache(std::vector<ReadRange> ranges) {
+    ranges = coalesceReadRanges(std::move(ranges), options_.holeSizeLimit, options_.rangeSizeLimit);
+
+    std::vector<RangeCacheEntry> newEntries = makeCacheEntries(ranges);
+    // Add new entries, themselves ordered by offset
+    if (entries_.size() > 0) {
+      std::vector<RangeCacheEntry> merged(entries_.size() + newEntries.size());
+      std::merge(entries_.begin(), entries_.end(), newEntries.begin(), newEntries.end(),
+                 merged.begin());
+      entries_ = std::move(merged);
+    } else {
+      entries_ = std::move(newEntries);
+    }
+  }
+
+  ReadRangeCache::BufferSlice ReadRangeCache::read(const ReadRange& range) {
+    if (range.length == 0) {
+      return {std::make_shared<Buffer>(*memoryPool_, 0), 0, 0};
+    }
+
+    const auto it = std::lower_bound(entries_.begin(), entries_.end(), range,
+                                     [](const RangeCacheEntry& entry, const ReadRange& range) {
+                                       return entry.range.offset + entry.range.length <
+                                              range.offset + range.length;
+                                     });
+
+    BufferSlice result{};
+    bool hit_cache = false;
+    if (it != entries_.end() && it->range.contains(range)) {
+      hit_cache = it->future.valid();
+      it->future.get();
```

Comment:
Should we catch and rethrow an orc::Exception?
Should we use timeout here to fallback to direct read?

Reply:
Maybe keeping the behavior consistency with `InputStream::read` is better? If `InputStream::read` throws an exception, it is transmitted directly to user.